File: src/solution/leetcode/45.cpp
#include "leetcode.h"
#include "utils.h"

class Solution {
 public:
  int jump(vector<int>& nums) {
    if (nums.size() == 1) return 0;
    vector<int> stepBoundary;
    stepBoundary.push_back(1);
    int idx = 0;
    for (int step = 0; step < stepBoundary.size(); step++) {
      int boundIdx = stepBoundary[step];
      int maxBound = -1;
      for (; idx < boundIdx; idx++) {
        maxBound = max(maxBound, idx + nums[idx] + 1);
      }
      if (maxBound >= nums.size()) return step + 1;
      stepBoundary.push_back(maxBound);
    }
    return -1;
  }
};

int main() {
  Solution sol;
  vector<int> arr = {2, 3, 1, 1, 4};
  // cout << sol.jump(arr) << endl;
  arr = {1, 1, 1, 1};
  cout << sol.jump(arr) << endl;
  return 0;
}

File: src/solution/leetcode/45.py
class Solution:

  def jump(self, nums: List[int]) -> int:
    jp = [9999] * len(nums)
    jp[len(nums) - 1] = 0
    for i in reversed(range(0, len(nums))):
      if i + 1 >= len(nums):
        continue
      for j in range(i + 1, min(len(nums), i + nums[i] + 1)):
        jp[i] = min(jp[i], jp[j] + 1)
    return jp[0]
